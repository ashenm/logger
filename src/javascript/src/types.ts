export enum Level {
  Info = "INFO",
  Warning = "WARNING",
  Debug = "DEBUG",
  Error = "ERROR",
}

export enum Environment {
  Development = "DEVELOPMENT",
  Staging = "STAGING",
  Production = "PRODUCTION",
}

export interface Context {
  raw: string;
}

export interface Constructor {
  environment?: Environment | string;
}

export interface Parameters {
  level: Level;
  message: string;
  context?: unknown;
}

export interface Message {
  level: Level;
  timestamp: string;
  message: string;
  context: Context;
  environment: string;
}
