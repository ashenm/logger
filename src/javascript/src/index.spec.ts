import Logger from "./index";
import { Environment } from "./types";
import { mockConsoleLog } from "../jest.setup";

describe("Logger.constructor", () => {
  // when
  beforeAll(() => {
    new Logger({ environment: "test-environment" }).info({ message: "test-message" });
  });

  it("Should use constructor overrides on explicit specify", () => {
    expect(mockConsoleLog.mock.calls).toMatchSnapshot();
  });

  afterAll(jest.clearAllMocks);
});

describe("Logger.constructor", () => {
  // given
  beforeAll(() => {
    Object.assign(process.env, { ENVIRONMENT: "test-env-environment" });
  });

  // when
  beforeAll(() => {
    new Logger().info({ message: "test-message" });
  });

  it("Should use environment variable overrides when available", () => {
    expect(mockConsoleLog.mock.calls).toMatchSnapshot();
  });

  afterAll(() => {
    Object.assign(process.env, { ENVIRONMENT: null });
    jest.clearAllMocks();
  });
});

describe("Logger.getInstance", () => {
  let logger: Logger;

  // given
  beforeAll(() => {
    logger = new Logger({ environment: Environment.Staging });
  });

  // then
  it("Should return singleton logger instance", () => {
    expect(Logger.getInstance()).toBe(logger);
  });

  afterAll(jest.clearAllMocks);
});

describe.each(["debug", "error", "info", "warn"])("Logger.prototype.%s", (level: string) => {
  const logger: Logger = Logger.getInstance();

  // when
  beforeAll(() => {
    logger[level]({ message: "test-message", context: "text-context" });
  });

  // then
  it(`Should log ${level} level entry with valid arity`, () => {
    expect(mockConsoleLog.mock.calls).toMatchSnapshot();
  });

  afterAll(jest.clearAllMocks);
});
