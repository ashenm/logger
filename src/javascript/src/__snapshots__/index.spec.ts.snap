// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logger.constructor Should use constructor overrides on explicit specify 1`] = `
[
  [
    "{"level":"INFO","environment":"test-environment","context":{},"message":"test-message","timestamp":"1970-01-01T00:00:00.000Z"}",
  ],
]
`;

exports[`Logger.constructor Should use environment variable overrides when available 1`] = `
[
  [
    "{"level":"INFO","environment":"test-env-environment","context":{},"message":"test-message","timestamp":"1970-01-01T00:00:00.000Z"}",
  ],
]
`;

exports[`Logger.prototype.debug Should log debug level entry with valid arity 1`] = `
[
  [
    "{"level":"DEBUG","environment":"DEVELOPMENT","context":{"raw":"\\"text-context\\""},"message":"test-message","timestamp":"1970-01-01T00:00:00.000Z"}",
  ],
]
`;

exports[`Logger.prototype.error Should log error level entry with valid arity 1`] = `
[
  [
    "{"level":"ERROR","environment":"DEVELOPMENT","context":{"raw":"\\"text-context\\""},"message":"test-message","timestamp":"1970-01-01T00:00:00.000Z"}",
  ],
]
`;

exports[`Logger.prototype.info Should log info level entry with valid arity 1`] = `
[
  [
    "{"level":"INFO","environment":"DEVELOPMENT","context":{"raw":"\\"text-context\\""},"message":"test-message","timestamp":"1970-01-01T00:00:00.000Z"}",
  ],
]
`;

exports[`Logger.prototype.warn Should log warn level entry with valid arity 1`] = `
[
  [
    "{"level":"WARNING","environment":"DEVELOPMENT","context":{"raw":"\\"text-context\\""},"message":"test-message","timestamp":"1970-01-01T00:00:00.000Z"}",
  ],
]
`;
